ImageInLib/ImageInLib/src/imageInterpolation.c

```diff
@@ -3,255 +3,200 @@
+    resultPoint.z = real_origin.z + real_spacing.sz * orientation.v3.z * image_point.z;
+    return resultPoint;
 }
 
-bool linear2dInterpolation(dataType** originalImage, dataType** newImage, size_t imageLength, size_t imageWidth, size_t imageHeight, dataType originalSpacing, dataType newSpacing)
-{
-    if (originalImage == NULL || newImage == NULL)
-        return false;
+Point3D getImageCoordFromRealCoord3D(Point3D real_point, Point3D real_origin, VoxelSpacing real_spacing, OrientationMatrix orientation) {
+    Point3D resultPoint;
+    resultPoint.x = (real_point.x - real_origin.x) / (real_spacing.sx * orientation.v1.x);
+    resultPoint.y = (real_point.y - real_origin.y) / (real_spacing.sy * orientation.v2.y);
+    resultPoint.z = (real_point.z - real_origin.z) / (real_spacing.sz * orientation.v3.z);
+    return resultPoint;
+}
 
-    size_t i, j, k, x, kn;
-    dataType k_int, k1, k2;
-    dataType divisionByOriginalSpacing = 1.0 / originalSpacing;
-
-    k_int = 0; kn = 0;
-    for (k = 0; k < imageHeight - 1; k++) {
-        k1 = k * originalSpacing;
-        k2 = k1 + originalSpacing;
-        do {
-            for (i = 0; i < imageLength; i++) {
-                for (j = 0; j < imageWidth; j++) {
-                    x = x_new(i, j, imageLength);
-                    newImage[kn][x] = (dataType)(originalImage[k][x] * ((k2 - k_int) * divisionByOriginalSpacing) +
-                        originalImage[k + 1][x] * ((k_int - k1) * divisionByOriginalSpacing));
-                }
-            }
-            k_int = k_int + newSpacing;
-            kn = kn + 1;
-        } while (k_int < k2);
-    }
+//=======================================================
+ 
+//2D images
+Point2D getRealCoordFromImageCoord2D(Point2D image_point, Point2D real_origin, PixelSpacing real_spacing, OrientationMatrix2D orientation) {
+    Point2D result_point;
+    result_point.x = real_origin.x + real_spacing.sx * orientation.v1.x * image_point.x;
+    result_point.y = real_origin.y + real_spacing.sy * orientation.v2.y * image_point.y;
+    return result_point;
+}
 
-    return true;
+Point2D getImageCoordFromRealCoord2D(Point2D real_point, Point2D real_origin, PixelSpacing real_spacing, OrientationMatrix2D orientation) {
+    Point2D result_point;
+    result_point.x = (real_point.x - real_origin.x) / (real_spacing.sx * orientation.v1.x);
+    result_point.y = (real_point.y - real_origin.y) / (real_spacing.sy * orientation.v2.y);
+    return result_point;
 }
 
-bool downSampling(dataType** originalImage, dataType** newImage, size_t length, size_t width, size_t height) {
+dataType getInterpolatedValueNearestNeighbor2D(Image_Data2D src_image, Point2D point) {
 
-    if (originalImage == NULL || newImage == NULL)
-        return false;
+    int i_floor, i_ceil, j_floor, j_ceil;
 
-    size_t lengthNew = (length / 2), widthNew = (width / 2), heightNew = (height / 2);
+    if (floor(point.x / src_image.spacing.sx) > src_image.height - 1) {
```

Comment:
I do not understand, why do you point coordinates by  spacing 
Point is in image CS (in other words pixel grid CS, where given value represents pixel index).
For this specific case (nearest neighbor) - you are searching for the value corresponding to the closest pixel center point so the pixel size is important, when you calculate distances, but it is not important here